### PR TITLE
feat(C-6): wire Hybrid Router into LangGraph workflow (#3182)

### LINKED ISSUE #3182: C-6: Graph Wiring for Hybrid Router
## Summary

Wire the Hybrid Router (C-2) into the LangGraph workflow graph. Currently, the Hybrid Router code is complete but not integrated into the actual workflow.

## Background

C-5 Pilot testing revealed that:
- `ENABLE_DYNAMIC_ROUTING=true` is set in staging
- Hybrid Router code exists in `core/flow/hybrid_router.py` and `core/flow/router_node.py`
- BUT: No `[ROUTER_FAST_PATH]` or `[ROUTER_SLOW_PATH]` logs appear
- Root cause: Hybrid Router is not wired into `langgraph_orchestrator.py`

## Current Graph Structure

```
reviewer → decision → hitl_gate → (fix/monitor_ci/publisher)
```

## Target Graph Structure (when ENABLE_DYNAMIC_ROUTING=true)

```
reviewer → router_node → hitl_gate → (fix/monitor_ci/publisher)
```

## Implementation Tasks

- [ ] Define `router_node` function in `langgraph_orchestrator.py`
- [ ] Add `router_node` to workflow graph
- [ ] Check `ENABLE_DYNAMIC_ROUTING` flag to decide routing path
- [ ] Wire `reviewer → router_node → hitl_gate` when flag is true
- [ ] Maintain backward compatibility (flag=false uses old decision logic)
- [ ] Add tests for the new routing path

## Acceptance Criteria

1. When `ENABLE_DYNAMIC_ROUTING=false`: Old behavior (reviewer → decision → hitl_gate)
2. When `ENABLE_DYNAMIC_ROUTING=true`: New behavior (reviewer → router_node → hitl_gate)
3. Logs show `[ROUTER_FAST_PATH]` or `[ROUTER_SLOW_PATH]` events when flag is true
4. All existing tests pass
5. CI passes

## Related Issues

- Parent: #3158 (C-5 Pilot)
- Depends on: #3149 (C-2 Hybrid Router)
- Depends on: #3160 (C-5 HITL Wiring)

## Labels

- `epic-c`
- `priority: P0`
- `type: feature`

### PR DESCRIPTION
**Pull request by RC918**
# feat(C-6): wire Hybrid Router into LangGraph workflow (#3182)

## Summary

This PR wires the Hybrid Router (C-2) into the LangGraph workflow graph. Previously, the Hybrid Router code existed in `core/flow/hybrid_router.py` but was never integrated into the actual workflow, causing `[ROUTER_FAST_PATH]` and `[ROUTER_SLOW_PATH]` events to never appear in logs.

**Changes:**
- Add `router_node` function that uses `HybridRoutingPolicy` for routing decisions
- Add router node to workflow graph
- Conditionally wire graph based on `ENABLE_DYNAMIC_ROUTING` flag:
  - When `true`: `reviewer → router → hitl_gate`
  - When `false`: `reviewer → decision → hitl_gate` (legacy behavior preserved)

**Graph Flow:**
```
ENABLE_DYNAMIC_ROUTING=true:  reviewer → router → hitl_gate → ...
ENABLE_DYNAMIC_ROUTING=false: reviewer → decision → hitl_gate → ... (legacy)
```

## Updates since last revision

- Fixed `test_graph_has_expected_nodes` test to include the new `router` node in expected nodes (21 nodes instead of 20)
- Added caching to `get_hybrid_router()` via `_CACHED_ROUTERS` dict to avoid expensive LLM initialization overhead on every call
- Added `"version": 1` field to `routing_decision` state dict for backward compatibility and future schema migrations
- Removed `decision.reasoning` from AIMessage content to prevent prompt injection risk (reasoning is still available in `state["routing_decision"]` for debugging)
- **Critical fix: Prevent infinite loop when `next_node == 'decision'`** - Force `merge_decision = "request_changes"` instead of using `state.get()` which could return `"pending"` and cause a loop: `ci_monitor → reviewer → router → hitl_gate → ci_monitor...`

## Review & Testing Checklist for Human

- [ ] **Verify infinite loop fix** (line 4833-4836) - When router returns `next_node='decision'` (HITL required), we now force `merge_decision='request_changes'` to prevent routing back to `monitor_ci`. Confirm this is the correct behavior for HITL cases.
- [ ] **Deploy to staging and verify logs show `[ROUTER_FAST_PATH]` or `[ROUTER_SLOW_PATH]` events** - This is the primary acceptance criteria. Trigger a test workflow and check logs.
- [ ] **Verify verdict mapping logic** (lines 4792-4800) - The mapping from old decision values (`needs_fix` → `request_changes`, `pending` → `unknown`) may have edge cases
- [ ] **Verify fallback behavior** (lines 4862-4888) - When HybridRouter fails, the fallback uses simplified decision logic. Confirm this matches expected behavior.
- [ ] **Confirm HITL gate works with router output** - The router sets `requires_hitl_approval` which hitl_gate should respect

### Test Plan
1. Merge this PR
2. Deploy to staging (worker will restart and pick up new graph wiring)
3. Trigger a test workflow via webhook endpoint
4. Check staging logs for `[ROUTER_FAST_PATH]` or `[ROUTER_SLOW_PATH]` events
5. Verify workflow completes successfully through hitl_gate
6. **Test HITL case**: Trigger a workflow that requires HITL approval and verify it doesn't loop

### Notes

- The router node is always added to the graph but only wired when `ENABLE_DYNAMIC_ROUTING=true`
- Graph structure is determined at compile time - requires worker restart if setting changes
- 127 flow controller tests + 14 HITL integration tests passed locally
- Code review comments addressed: caching, versioning, prompt injection prevention, and infinite loop fix

Closes #3182

---
Link to Devin run: https://app.devin.ai/sessions/4c624f57c50e499f98859c62908c800d
Requested by: Ryan Chen (ryan2939z@gmail.com) / @RC918

### CODE REVIEW

File: handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py [hunk 1]
```diff
@@ -4743,6 +4743,155 @@ def decision_node(state: AgentState) -> AgentState:
...
+            merge_decision = "request_changes"
+
+        state["merge_decision"] = merge_decision
+        state["requires_hitl_approval"] = requires_hitl
+        state["routing_decision"] = {
+            "next_node": next_node,
+            "reasoning": decision.reasoning,
+            "risk_assessment": decision.risk_assessment,
+            "requires_hitl_approval": requires_hitl,
+        }
```
Comment (RC918) [inline handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py]: IMPACT: This introduces a new unversioned structured field (`routing_decision`) into the agent state without backward compatibility safeguards or schema versioning (e.g., no `routing_decision_v2` or `diagnostic_version` bump). Downstream systems (logging, monitoring, runbooks) may break if they expect a flat structure or lack error tolerance. SUGGESTED FIX: Add a version key (e.g., `version: 'v1'`) inside `routing_decision` and document it in the docstring. Alternatively, use a version-suffixed key like `routing_decision_v1`.

File: handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py [hunk 2]
```diff
@@ -4743,6 +4743,155 @@ def decision_node(state: AgentState) -> AgentState:
...
+        "operation": "router",
+        "trace_id": trace_id,
+        "verdict": verdict,
+        "severity": severity,
+        "blocker_count": blocker_count,
+    })
+
+    try:
+        # Get Hybrid Router instance (with LLM for slow path)
+        router = get_hybrid_router(use_llm=True)
```
Comment (RC918) [inline handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py]: IMPACT: Creating a new router instance on every call with `use_llm=True` may lead to expensive LLM initialization overhead per invocation. If the router or its LLM client is not internally cached/singleton, this can severely impact latency and cost. SUGGESTED FIX: Ensure `get_hybrid_router` caches instances by `use_llm` flag or inject the router as a dependency instead of constructing it inline.

File: handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py [hunk 3]
```diff
@@ -4743,6 +4743,155 @@ def decision_node(state: AgentState) -> AgentState:
...
+            "operation": "router",
+            "trace_id": trace_id,
+            "next_node": next_node,
+            "merge_decision": merge_decision,
+            "requires_hitl_approval": requires_hitl,
+            "reasoning": decision.reasoning,
+        })
+
+        state["messages"] = state.get("messages", []) + [
+            AIMessage(content=f"Router decision: {next_node}. Reason: {decision.reasoning}")
```
Comment (RC918) [inline handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py]: IMPACT: Direct string interpolation of `decision.reasoning` into an AI message risks exposing sensitive data or prompt injection vectors if later used in LLM context. Unsanitized reasoning text could include PII or malicious payloads. SUGGESTED FIX: Sanitize or truncate `decision.reasoning` before embedding in messages, or use structured metadata instead of plain text content.

Comment (RC918) [reply handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py]: Fixed in commit 91a8c3dc. Removed `decision.reasoning` from the AIMessage content to prevent prompt injection risk. The reasoning is still available in `state["routing_decision"]` for debugging purposes.

File: handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py (hunk 2 above)
Comment (RC918) [reply handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py]: Fixed in commit 91a8c3dc. Added module-level caching (`_CACHED_ROUTERS`) to `get_hybrid_router()` keyed by `use_llm` flag. Custom `llm_generate_fn` bypasses the cache for flexibility in testing.

File: handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py (hunk 1 above)
Comment (RC918) [reply handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py]: Fixed in commit 91a8c3dc. Added `"version": 1` field to `routing_decision` dict for backward compatibility and future schema migrations.
